src/runtime/api.rs

```diff
@@ -197,13 +197,14 @@ pub struct Session<'a> {
 
 // Session is thread-compatible.
 unsafe impl Send for Session<'_> {}
+unsafe impl Sync for Session<'_> {}
```

Comment:
https://github.com/iree-org/iree/blob/e8f4948d5c62beea27a109161b697d9088d5b3bd/runtime/src/iree/runtime/session.h#L22-L42
The docs says that caller must use external synchronization if they will be using it or any resource derived from it concurrently, so Session is Send but not Sync?

Reply:
Now that you ask, I double checked and the way the Session impls are defined this is actually not safe.
This is because they take a const reference to self and still call methods that require a mut session pointer, so if it is sync different threads could use their const reference and call simultaneously.

If we change the methods in question to take `&mut self` I think this should be fine then, the borrow checker should make sure all calls are correctly synchronized. That is, if we don't allow any other options for users to get their hands on a separate Session object pointing to the same iree session (eg by implementing clone as an increment of the reference counter, or through access methods of other objects).

I guess there just are several different ways of wrapping this in a safe way, I'll remove this from the PR for now, but maybe this is worth investigating a bit more? (I guess this is also related to the question if Session should have a a lifetime parameter on the instance or just increment its refcount).

Reply:
I agree, but before taking `&mut Session`, removing the session/instance lifetime from structs like `vm::List`, `vm::Function`, and `hal::BufferView` is necessary. Using const references was to bypass the lifetime requirement through interior mutability. But since session/instance lifetime is unnecessary due to atomic refcount, this can be fixed through #16.

I might be slow to fix this since I am on my mandatory military duty though. What is your use case, is there anything I can help with?